### PR TITLE
feat: add unit tests for reconstructAllDataColumnSidecars method

### LINKED ISSUE #9904: Improve `MiscHelpersFulu` unit test coverage
for `reconstructAllDataColumnSidecars`, `WithPrecomputeBenchmark` can be used as reference

### PR DESCRIPTION
**Pull request by DeVikingMark**
## Add unit tests for reconstructAllDataColumnSidecars method

**Tests added:**
- Success case with sufficient sidecars
- Error handling for insufficient sidecars  
- Empty collection handling
- Edge case with exactly half sidecars
- Sidecar ordering preservation
- Single sidecar error case

## Fixed Issue(s)
fixes #9904

## Documentation

- [ ] I thought about documentation and added the `doc-change-required` label to this PR if updates are required.

## Changelog

- [ ] I thought about adding a changelog entry, and added one if I deemed necessary.

---

> [!NOTE]
> Adds comprehensive tests for `reconstructAllDataColumnSidecars` with shared fixtures and initializes KZG for realistic verification.
> 
> - **Tests (Fulu helpers)**:
>   - Add coverage for `reconstructAllDataColumnSidecars`:
>     - Succeeds with sufficient or exactly half sidecars
>     - Fails with insufficient or empty input (throws `IllegalArgumentException`)
>     - Preserves sidecar ordering
>     - Works with all sidecars
> - **Test setup**:
>   - Introduce static `SPEC` and initialize KZG via `TrustedSetupLoader`; reinitialize spec for testing
>   - Add `@BeforeAll` to create shared blobs, `SignedBeaconBlock`, and original sidecars used across tests
>   - Update references from `spec` to `SPEC` and adjust helpers acquisition accordingly
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit b2763510ca95c1906abb227d1fdc68c3f4ee1376. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

**Comment by DeVikingMark** [comment]
I have read the CLA Document and I hereby sign the CLA

File: ethereum/spec/src/test/java/tech/pegasys/teku/spec/logic/versions/fulu/helpers/MiscHelpersFuluTest.java [hunk 1]
```diff
@@ -325,4 +329,171 @@ static Stream<Arguments> getValidatorCustodyRequirementFixtures() {
...
+        originalSidecars.subList(halfSize, originalSidecars.size());
+
+    final List<DataColumnSidecar> reconstructedSidecars =
+        miscHelpersFulu.reconstructAllDataColumnSidecars(partialSidecars, kzg);
+
+    // Verify that all sidecars were reconstructed
+    assertThat(reconstructedSidecars).hasSize(originalSidecars.size());
+
+    // Verify that the reconstructed sidecars have the same structure as original
+    for (int i = 0; i < reconstructedSidecars.size(); i++) {
```
Comment (zilm13) [inline ethereum/spec/src/test/java/tech/pegasys/teku/spec/logic/versions/fulu/helpers/MiscHelpersFuluTest.java]: assertThat(reconstructedSidecars).isEqualTo(originalSidecars);
would be fine

File: ethereum/spec/src/test/java/tech/pegasys/teku/spec/logic/versions/fulu/helpers/MiscHelpersFuluTest.java [hunk 2]
```diff
@@ -325,4 +329,171 @@ static Stream<Arguments> getValidatorCustodyRequirementFixtures() {
...
+  @Test
+  public void reconstructAllDataColumnSidecars_withSufficientSidecars_shouldReconstructAll() {
+    // Create test data similar to SidecarBenchmarkConfig
+    final List<Blob> blobs =
+        IntStream.range(0, 4).mapToObj(__ -> dataStructureUtil.randomValidBlob()).toList();
+
+    final KZG kzg = KZG.getInstance(false);
+    TrustedSetupLoader.loadTrustedSetupForTests(kzg);
+    final List<List<MatrixEntry>> extendedMatrix =
+        miscHelpersFulu.computeExtendedMatrixAndProofs(blobs, kzg);
```
Comment (zilm13) [inline ethereum/spec/src/test/java/tech/pegasys/teku/spec/logic/versions/fulu/helpers/MiscHelpersFuluTest.java]: I see original issue, but we'd better use `constructDataColumnSidecarsOld` for this. 
Though it's marked as `@Deprecated` which could be a bit confusing, it's just an over-protection to avoid using it anywhere but in the tests

File: ethereum/spec/src/test/java/tech/pegasys/teku/spec/logic/versions/fulu/helpers/MiscHelpersFuluTest.java [hunk 3]
```diff
@@ -325,4 +329,171 @@ static Stream<Arguments> getValidatorCustodyRequirementFixtures() {
...
+        dataStructureUtil.randomSignedBeaconBlockWithCommitments(blobs.size());
+
+    final List<DataColumnSidecar> originalSidecars =
+        miscHelpersFulu.constructDataColumnSidecars(
+            signedBeaconBlock.getMessage(), signedBeaconBlock.asHeader(), extendedMatrix);
+
+    // Test with more than half of the sidecars (should succeed)
+    final int halfSize = originalSidecars.size() / 2;
+    final List<DataColumnSidecar> partialSidecars =
+        originalSidecars.subList(halfSize, originalSidecars.size());
```
[the remark this answers is not in the document]
Comment (zilm13) [reply ethereum/spec/src/test/java/tech/pegasys/teku/spec/logic/versions/fulu/helpers/MiscHelpersFuluTest.java]: @DeVikingMark Cursor is right, having exactly half of the sidecars here there is no difference compared to `reconstructAllDataColumnSidecars_withExactlyHalfSidecars_shouldSucceed`. Start with halfsize -1, for example to fix this

File: ethereum/spec/src/test/java/tech/pegasys/teku/spec/logic/versions/fulu/helpers/MiscHelpersFuluTest.java [hunk 4]
```diff
@@ -325,4 +329,171 @@ static Stream<Arguments> getValidatorCustodyRequirementFixtures() {
...
+    final List<List<MatrixEntry>> extendedMatrix =
+        miscHelpersFulu.computeExtendedMatrixAndProofs(blobs, kzg);
+    final SignedBeaconBlock signedBeaconBlock =
+        dataStructureUtil.randomSignedBeaconBlockWithCommitments(blobs.size());
+
+    final List<DataColumnSidecar> originalSidecars =
+        miscHelpersFulu.constructDataColumnSidecars(
+            signedBeaconBlock.getMessage(), signedBeaconBlock.asHeader(), extendedMatrix);
+
+    // Test with more than half of the sidecars
```
Comment (zilm13) [inline ethereum/spec/src/test/java/tech/pegasys/teku/spec/logic/versions/fulu/helpers/MiscHelpersFuluTest.java:416]: it's exactly a half, confusing comment

File: ethereum/spec/src/test/java/tech/pegasys/teku/spec/logic/versions/fulu/helpers/MiscHelpersFuluTest.java [hunk 5]
```diff
@@ -325,4 +329,171 @@ static Stream<Arguments> getValidatorCustodyRequirementFixtures() {
...
+    final int halfSize = originalSidecars.size() / 2;
+    final List<DataColumnSidecar> partialSidecars =
+        originalSidecars.subList(halfSize, originalSidecars.size());
+
+    final List<DataColumnSidecar> reconstructedSidecars =
+        miscHelpersFulu.reconstructAllDataColumnSidecars(partialSidecars, kzg);
+
+    // Verify that sidecars are ordered by index
+    for (int i = 0; i < reconstructedSidecars.size() - 1; i++) {
+      assertThat(reconstructedSidecars.get(i).getIndex())
```
Comment (zilm13) [inline ethereum/spec/src/test/java/tech/pegasys/teku/spec/logic/versions/fulu/helpers/MiscHelpersFuluTest.java]: I'd verify that it `i == getIndex` instead

File: ethereum/spec/src/test/java/tech/pegasys/teku/spec/logic/versions/fulu/helpers/MiscHelpersFuluTest.java [hunk 6]
```diff
@@ -325,4 +329,171 @@ static Stream<Arguments> getValidatorCustodyRequirementFixtures() {
...
+          .isLessThanOrEqualTo(reconstructedSidecars.get(i + 1).getIndex());
+    }
+  }
+
+  @Test
+  public void reconstructAllDataColumnSidecars_withSingleSidecar_shouldThrowException() {
```
Comment (zilm13) [inline ethereum/spec/src/test/java/tech/pegasys/teku/spec/logic/versions/fulu/helpers/MiscHelpersFuluTest.java]: I think case with less than half is enough, we don't need this one. Instead I'd add case with all sidecars, I have no idea actually what will happen.

File: ethereum/spec/src/test/java/tech/pegasys/teku/spec/logic/versions/fulu/helpers/MiscHelpersFuluTest.java [hunk 7]
```diff
@@ -325,4 +329,171 @@ static Stream<Arguments> getValidatorCustodyRequirementFixtures() {
...
+        IntStream.range(0, 4).mapToObj(__ -> dataStructureUtil.randomValidBlob()).toList();
+
+    final KZG kzg = KZG.getInstance(false);
+    TrustedSetupLoader.loadTrustedSetupForTests(kzg);
+    final List<List<MatrixEntry>> extendedMatrix =
+        miscHelpersFulu.computeExtendedMatrixAndProofs(blobs, kzg);
+    final SignedBeaconBlock signedBeaconBlock =
+        dataStructureUtil.randomSignedBeaconBlockWithCommitments(blobs.size());
+
+    final List<DataColumnSidecar> originalSidecars =
```
Comment (zilm13) [inline ethereum/spec/src/test/java/tech/pegasys/teku/spec/logic/versions/fulu/helpers/MiscHelpersFuluTest.java]: Looks like we could use same originalSidecars in all cases except one with empty list  + creation sidecars is expensive, so I'd put it in @BeforeAll block to avoid running this computation for 5 times.

**Review by zilm13 — COMMENTED** [review]
Hi!
Thank you for your contribution. I've left some comments which is good to address.

**Review by zilm13 — APPROVED** [review]
LGTM
just merge the latest master and fix the build 
kzg is now a part of the spec and you don't need to have it separately

**Comment by zilm13** [comment]
you don't need to pass kzg separately anymore, see https://github.com/Consensys/teku/blob/28b6cabb2f2a3cce3844bd9a73fa8217de45528d/storage/src/test/java/tech/pegasys/teku/storage/client/BlobSidecarReconstructionProviderTest.java#L206-L211 for example

**Comment by zilm13** [comment]
but you need to reinitilize spec with real kzg (in BeforeAll)
```
    final KZG kzg = KzgRetriever.getKzgWithLoadedTrustedSetup("minimal");
    spec.reinitializeForTesting(        AvailabilityCheckerFactory.NOOP_BLOB_SIDECAR,
        AvailabilityCheckerFactory.NOOP_DATACOLUMN_SIDECAR,, kzg);



```

**Comment by zilm13** [comment]
If you are stuck, mention me, I will fix it

**Comment by zilm13** [comment]
Looks good, merging. @DeVikingMark thank you for your help!

**Comment by DeVikingMark** [comment]
@zilm13 thanks for your support! :)
